### ISSUE TITLE
Yolo8/11 classify task (libtorch, C++)

### ISSUE THREAD

**Issue by yuriiluzhkov** [ultralytics/ultralytics#17734]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I have a task - to implement Yolo8/11 classification task using C++ libtorch (not detection task). Unfortunately I have no example code, could you please help me with some example?

this [example ](https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-LibTorch-CPP-Inference) works properly for detection task and its impossible to use it for classification

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yuriiluzhkov, thanks for reaching out to Ultralytics 🚀! We appreciate your interest in using YOLO for classification with libtorch.

For new users, we recommend taking a look at our [Docs](https://docs.ultralytics.com), where you can find extensive information on both [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage. You might discover that some of your questions are already covered there.

Regarding your query on using libtorch for classification tasks, if you are experiencing specific issues or have a bug report 🐛, providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) will greatly assist us in helping you debug the problem.

For questions related to custom implementations or training ❓, please share detailed information about your task, including any dataset samples or logs you might have. Ensure you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Feel free to engage with our community through multiple channels:
- Join us on [Discord](https://ultralytics.com/discord) for real-time chat 🎧
- Dive into discussions on [Discourse](https://community.ultralytics.com)
- Participate in threads on our [Subreddit](https://reddit.com/r/ultralytics)

## Upgrade

Please make sure you are using the latest version of the `ultralytics` package by upgrading it along with all required dependencies in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

Run YOLO in the following compatible environments (all necessary dependencies, including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) are preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

Currently, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing, ensuring correct operations of YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu.

This automated response is here to guide you while an Ultralytics engineer will take a closer look shortly.

**Comment by Y-T-G** [comment]
For classification, the code would be much simpler. You will have to use this preprocessing instead of letterbox.

https://github.com/ultralytics/ultralytics/blob/426879d80d49d0180b525c4fc2484772f9f6f8cc/ultralytics/data/augment.py#L2341

And for post-processing, you just get the max class score and index of the output.

**Comment by yuriiluzhkov** [comment]
> For classification, the code would be much simpler. You will have to use this preprocessing instead of letterbox.
> 
> https://github.com/ultralytics/ultralytics/blob/426879d80d49d0180b525c4fc2484772f9f6f8cc/ultralytics/data/augment.py#L2341
> 
> And for post-processing, you just get the max value and argmax index of the output.

Thank you for your answer, could you telll how this preprocesing can be called in case of c++ (libtorch)?

**Comment by glenn-jocher** [comment]
@yuriiluzhkov to replicate the preprocessing in C++ (LibTorch), you'll need to resize the input image to the required dimensions (e.g., 224x224 for classification models) and normalize the pixel values (e.g., divide by 255 and standardize using mean and standard deviation). LibTorch provides functions like `torch::from_blob` to convert image data into tensors and operations for resizing and normalizing.

Here's a minimal example for preprocessing in C++ using LibTorch:

```cpp
#include <torch/torch.h>
#include <opencv2/opencv.hpp>

// Function to preprocess the image
torch::Tensor preprocess(const cv::Mat& image) {
    // Resize to desired dimensions (e.g., 224x224)
    cv::Mat resized_image;
    cv::resize(image, resized_image, cv::Size(224, 224));

    // Convert to float tensor and normalize
    torch::Tensor img_tensor = torch::from_blob(resized_image.data, {1, resized_image.rows, resized_image.cols, 3}, torch::kUInt8);
    img_tensor = img_tensor.permute({0, 3, 1, 2}).to(torch::kFloat32) / 255.0; // Change to CxHxW and normalize to [0, 1]

    // Standardize using mean and std (example values for RGB normalization)
    img_tensor = torch::nn::functional::normalize(img_tensor, /* mean */ {0.485, 0.456, 0.406}, /* std */ {0.229, 0.224, 0.225});
    return img_tensor;
}
```

You can call this method to preprocess your input image before running it through the LibTorch inference model. If you have further questions, feel free to ask!

**Comment by yuriiluzhkov** [comment]
> @yuriiluzhkov to replicate the preprocessing in C++ (LibTorch), you'll need to resize the input image to the required dimensions (e.g., 224x224 for classification models) and normalize the pixel values (e.g., divide by 255 and standardize using mean and standard deviation). LibTorch provides functions like `torch::from_blob` to convert image data into tensors and operations for resizing and normalizing.
> 
> Here's a minimal example for preprocessing in C++ using LibTorch:
> 
> ```c++
> #include <torch/torch.h>
> #include <opencv2/opencv.hpp>
> 
> // Function to preprocess the image
> torch::Tensor preprocess(const cv::Mat& image) {
>     // Resize to desired dimensions (e.g., 224x224)
>     cv::Mat resized_image;
>     cv::resize(image, resized_image, cv::Size(224, 224));
> 
>     // Convert to float tensor and normalize
>     torch::Tensor img_tensor = torch::from_blob(resized_image.data, {1, resized_image.rows, resized_image.cols, 3}, torch::kUInt8);
>     img_tensor = img_tensor.permute({0, 3, 1, 2}).to(torch::kFloat32) / 255.0; // Change to CxHxW and normalize to [0, 1]
> 
>     // Standardize using mean and std (example values for RGB normalization)
>     img_tensor = torch::nn::functional::normalize(img_tensor, /* mean */ {0.485, 0.456, 0.406}, /* std */ {0.229, 0.224, 0.225});
>     return img_tensor;
> }
> ```
> 
> You can call this method to preprocess your input image before running it through the LibTorch inference model. If you have further questions, feel free to ask!

Thank you for your answer, could you tell please if const cv::Mat& image assumes to be in RGB or BGR format? It seems to me that it is the reason of my incorrect results

**Comment by glenn-jocher** [comment]
Great question! The input `cv::Mat` assumes a BGR format because OpenCV typically reads images in that format by default. Since the model expects RGB, you need to convert the image using `cv::cvtColor(image, image, cv::COLOR_BGR2RGB)` before further preprocessing. I hope this helps clarify your results!

**Comment by yuriiluzhkov** [comment]
> Great question! The input `cv::Mat` assumes a BGR format because OpenCV typically reads images in that format by default. Since the model expects RGB, you need to convert the image using `cv::cvtColor(image, image, cv::COLOR_BGR2RGB)` before further preprocessing. I hope this helps clarify your results!

May be this example contains a little bug? https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-LibTorch-CPP-Inference/main.cc#L226
there is no any color conversion BGR2RGB before processing in model

**Comment by glenn-jocher** [comment]
You're correct; the example could be missing a BGR-to-RGB color conversion before passing the image to the model, as models typically expect RGB input. This step is critical if you're using OpenCV to read images (default BGR format). Thank you for pointing this out; we'll review and address it if necessary. You can apply the conversion manually using `cv::cvtColor(image, image, cv::COLOR_BGR2RGB)` for now.
